### ISSUE TITLE
distribute PEP 561 stubs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9717]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#9717|
|Type|enhancement|
|Created|2019-10-17 23:10:58Z|

We already have lots of information in pydoctor annotations that we could possibly extract and teach mypy about.  (Maybe even with a plugin, so we don't have to do a giantic source-to-source rewrite commit?)

Anyway, most interestingly here, `Deferred`  should be made into a generic, so we can have `Deferred[X]` to indicate a `Deferred` that fires with `X`.

<details><summary>Searchable metadata</summary>

```
trac-id__9717 9717
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__publish_types publish types
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__None None
time__1571353858118476 1571353858118476
changetime__1594360403322407 1594360403322407
version__None None
owner__None None
cc__mthuurne cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_1"></a>|@mthuurne commented|
|-|-|

In the long term, I think it makes sense to teach pydoctor about type annotations and then move the type info from the docstrings to annotations. This approach makes pydoctor more useful for projects other than Twisted. So in my eyes, that would make extracting the type info from the docstrings more attractive than writing a mypy plugin and keeping the type info in the docstrings.

You wrote PEP 561 stubs in the ticket title, which would imply external stubs. But if it is the Twisted project maintaining the stubs, putting the annotations inside the code base itself would also be an option. In general, keeping code and information about code close together is a good idea (see docstrings), but perhaps this should be a long-term goal and not the way to start.

Which Python version should the type stubs target? Twisted supports Python 2.7 and Python 3.5+ at the moment, but developers stuck on Python 2.7 are probably not all that interested in static type checking, since it requires a big time investment. If they are in a position to put that much time into upgrading their code base, they're probably better off porting to Python 3 first.

The main difference between Python 3.5 and 3.6 is the annotation syntax for variables/fields. The 3.6 syntax is a lot more elegant, but the 3.5 syntax is not too terrible. Still, if Twisted is going to move to 3.6 as its minimum required version soonish, we could just use that for the stubs from the start.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

> In the long term, I think it makes sense to teach pydoctor about type annotations and then move the type info from the docstrings to annotations. This approach makes pydoctor more useful for projects other than Twisted. So in my eyes, that would make extracting the type info from the docstrings more attractive than writing a mypy plugin and keeping the type info in the docstrings.

I definitely agree, with the caveat being "in the long term".  We have tons and tons of type information in the code right now, and I'd love to be able to use it for developing Twisted itself and developing with Twisted without having to generate a 5 million line diff somehow - certainly not if we have to do it manually ;).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

> Which Python version should the type stubs target? Twisted supports Python 2.7 and Python 3.5+ at the moment, but developers stuck on Python 2.7 are probably not all that interested in static type checking, since it requires a big time investment. If they are in a position to put that much time into upgrading their code base, they're probably better off porting to Python 3 first.

On the contrary - one of the biggest users of Python 3 (Dropbox) specifically used MyPy on Py2 to improve the reliability of the code _first_, so they _could_ move to py3!  So I'd love for our last release that supports py2 to have those type stubs.

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_4"></a>|@mthuurne commented|
|-|-|

Replying to [Glyph](#note_2):
> > In the long term, I think it makes sense to teach pydoctor about type annotations and then move the type info from the docstrings to annotations. This approach makes pydoctor more useful for projects other than Twisted. So in my eyes, that would make extracting the type info from the docstrings more attractive than writing a mypy plugin and keeping the type info in the docstrings.
> 
> I definitely agree, with the caveat being "in the long term".  We have tons and tons of type information in the code right now, and I'd love to be able to use it for developing Twisted itself and developing with Twisted without having to generate a 5 million line diff somehow - certainly not if we have to do it manually ;).

A quick grep for "@type" has 8.4k hits, so a diff would be large but still several orders of magnitude below 5 million ;) For context: this is a bit over 2% of the combined line count of all `.py` files under `src/`.

Personally, I don't mind a big diff once or twice. Since we agree that PEP 484 annotations in the source is the right long term solution, we'll have to do one big diff sooner or later anyway.


Replying to [Glyph](#note_3):
> > Which Python version should the type stubs target? Twisted supports Python 2.7 and Python 3.5+ at the moment, but developers stuck on Python 2.7 are probably not all that interested in static type checking, since it requires a big time investment. If they are in a position to put that much time into upgrading their code base, they're probably better off porting to Python 3 first.
> 
> On the contrary - one of the biggest users of Python 3 (Dropbox) specifically used MyPy on Py2 to improve the reliability of the code _first_, so they _could_ move to py3!  So I'd love for our last release that supports py2 to have those type stubs.

For Python 2 support, putting the annotations in the source is only possible with `# type:` style comments, which is really ugly, in my opinion. So that would strengthen the case for a different short-term solution.


Getting the current type info into mypy should absolutely be automated. However, the current "@type" fields in the docstrings are not a formally defined language, so it will require some manual intervention in a minority of cases where non-trivial English is used. The question is where such interventions should be maintained.

In general, I think the options are to maintain them in the input (docstrings), in the converter (extractor or mypy plugin) or in the output (stubs, if any).

Maintaining them in the input, that is updating the docstring to use a pattern that the converter can parse, seems elegant at a first glance. However, this effectively means designing a second type annotation language that has the same functionality as PEP 484 and Twisted developers having to learn two languages. It would be an investment that doesn't take us closer to the long-term goal.

An alternative way to maintain them in the input would be to use PEP 484 syntax but keep the type info inside the docstrings until Python 2 support is dropped. This would require a big diff, writing a mypy plugin and updating pydoctor to understand PEP 484 syntax.

A down side to maintaining the corrections in the input is that getting updates out to the public is tied to Twisted's release cycle. I imagine we'd want to have a faster release cycle since this is experimental work.

Maintaining a list of exceptions in the converter would become a mess, I think. It would have the same drawback of having a second annotation language, but with a more complex language.

Applying interventions to the output is a clean and simple process, but it requires the conversion to be a one-time thing. So this means the type info would be extracted into PEP 561 style stubs, which are then used by both mypy and pydoctor. As soon as pydoctor can get the type info from stubs, the `@type` tags should be removed from the docstrings.

A benefit of stubs over a mypy plugin is that it doesn't favor any particular tool. While mypy is the de facto standard at the moment, in general I think it's a good idea to support formats rather than implementations.

So my preferred approach would be to extract PEP 561 style stubs using an automated extractor and manually correct in the stubs the minority of cases that are too exotic to easily automate. Release this as a separate stubs package on PyPI. After Twisted drops Python 2 support, apply the stubs in the Twisted source tree; automated tools for this already exist.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

> So my preferred approach would be to extract PEP 561 style stubs using an automated extractor and manually correct in the stubs the minority of cases that are too exotic to easily automate. Release this as a separate stubs package on PyPI. After Twisted drops Python 2 support, apply the stubs in the Twisted source tree; automated tools for this already exist.

This all sounds fairly reasonable, except for the part where I think there's going to be quite a lot of manual corrections, as we converge on `L{Deferred[X]}` rather than a thousand variant phrasings of `L{Deferred} that fires with L{X}` :).  In order to figure out what those manual corrections are though, surely we'd need mypy to run over Twisted itself?

Also - can you point towards the tool for automatically applying stubs to the codebase? that sounds interesting!

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_7"></a>|@mthuurne commented|
|-|-|

Replying to [Glyph](#note_6):
> > So my preferred approach would be to extract PEP 561 style stubs using an automated extractor and manually correct in the stubs the minority of cases that are too exotic to easily automate. Release this as a separate stubs package on PyPI. After Twisted drops Python 2 support, apply the stubs in the Twisted source tree; automated tools for this already exist.
> 
> This all sounds fairly reasonable, except for the part where I think there's going to be quite a lot of manual corrections, as we converge on `L{Deferred[X]}` rather than a thousand variant phrasings of `L{Deferred} that fires with L{X}` :).  In order to figure out what those manual corrections are though, surely we'd need mypy to run over Twisted itself?

I assumed that you can check Twisted itself against the work-in-progress PEP 561 stubs. If that is not possible out of the box, we could always modify mypy to make it possible.

> Also - can you point towards the tool for automatically applying stubs to the codebase? that sounds interesting!

The tool is called "retype": https://pypi.org/project/retype/

One of the MonkeyType developers told me about it after I modified MonkeyType to apply stubs and suggested that as a feature: https://github.com/Instagram/MonkeyType/issues/140

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_8"></a>|@mthuurne commented|
|-|-|

So far I've come across the following ways of generating type stubs:

* from the @type fields in docstrings (using a modified pydoctor?)
* MonkeyType (collects runtime type info)
* mypy's stubgen
* pytype (type inference)

Since I don't expect any of these approaches to be superior in all cases, maybe a good strategy is to use all of them and then write a tool to compare the generated stubs. If they all produce the same output for a particular function or field, its type is probably trivial. If they disagree, manual inspection is required to pick the best annotation.

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_9"></a>|@mthuurne commented|
|-|-|

I made an overview of the type annotation status of Twisted's dependencies:

[https://gist.github.com/mthuurne/0022865c96ff6eda0c7b13bf396f4b89]

The bad news is that most of these packages have no type annotations or stubs yet.

The good news is that Twisted's public interface uses very few types from these packages. So while the lack of stubs limits how useful mypy is for checking Twisted itself, it's not a blocker for creating type stubs for Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_10"></a>|@wsanchez commented|
|-|-|

Type annotations in hyperlink are in a PR awaiting review: https://github.com/python-hyper/hyperlink/pull/82

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_11"></a>|@wsanchez commented|
|-|-|

If **all** that happened in a first round was to make `Deferred[X]` work, that would be a big win for clients, since that would enable them to write type hints for things that return a `Deferred`.

Personally, I think the best approach is an incremental manual move to type hints.

While Glyph is right that it is more thorough to do the type hints before killing Python 2.7, I think that's actually more work than it's worth.  That would have been great as part of porting to Py3, but Twisted already supports Py3.  I'd go straight to Py3.6, or 3.5 if it's going to remain supported for a while yet.

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_13"></a>|@mthuurne commented|
|-|-|

Thanks for the Hyperlink PR; I updated the Gist.

I agree that an incremental approach is the way to go. I attempted the opposite approach this weekend: generating stubs for all of Twisted. And the results weren't very useful:

- `stubgen` (included with mypy) quickly produced a full set of `.pyi` files, but the generated stubs were mostly unannotated or using `Any`; I think it only got function arguments with defaults correct
- `pytype` generated more meaningful stubs, but it stopped because of errors at around 20% in; this was while using the `-k` (keep going) option

I think there are a couple of areas that are good to start with:

- `Deferred`, as you mentioned
- the zope interfaces, since those both imply the types for implementations and define the way components talk to each other
- extracting type information from docstrings, since those reflect the types that the developer intended

Regarding Python 2 vs 3, I originally mixed up two separate issues:

First issue is the syntax used for the type hints. Inside the source code, we can only use the full annotation syntax on Python 3.6+. But `.pyi` files are always in the "3.6" annotation syntax regardless of the Python version used, since these are only parsed by the type checker tool (mypy etc) and not by Python itself.

Second issue is the Python version that we're type-checking for. This is mainly important for strings, since byte and Unicode strings are somewhat interchangeable in Python 2 and entirely separate in Python 3. Since much of Twisted is network protocols, it has to deal with both: bytes going over the wire and strings used at the application level.

Some functions/methods in Twisted accept both byte and Unicode strings in Python 3. Is that an accidental feature that arose from converting the code from Python 2 or is that something that is desired? In the former case, perhaps we can clean up the API by picking one string type for the type hints and deprecate the other.

Another question is in which repository the type hints should live. I think we all agree that the desired long-term solution is having type annotations in Python 3.6 syntax inside the Twisted source code. But the question is what we do in the short term:

- inside the Twisted source code; in this case we must use `# type` comments to not break Python 2.7's parser
- as `.pyi` files in Twisted's repository
- as `.pyi` files in typeshed's repository
- as `.pyi` files in a dedicated `twisted-stubs` repository

My preference would be a dedicated `twisted-stubs` repository, since that would allow for the quickest iterations. Having to go through code reviews for every change is good for mature projects, where minimizing the risk of breaking functionality people depend upon is more important than enabling quick development. But since the type hints will start very small and need to grow a lot over time, I think a light weight process is better for the time being.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

I don't know if I've mentioned that we already have some in-house stubs at work for Twisted which do `Deferred[X]` and I really ought to open source it so we can absorb it in some form.

Given that we're <1 month away from the end of 2.7 support, possibly it's not worth investing a ton in it :-).

> My preference would be a dedicated twisted-stubs repository, since that would allow for the quickest iterations. Having to go through code reviews for every change is good for mature projects, where minimizing the risk of breaking functionality people depend upon is more important than enabling quick development. But since the type hints will start very small and need to grow a lot over time, I think a light weight process is better for the time being.

On this, I disagree.  The stubs-as-separate-repository setup of many `mypy` projects is a huge drag on new features being added and correctly type-checked.

But I can see that we wouldn't want to code-review every single function signature from scratch.

I think we can address both of these issues by getting to a setup where we generate `.pyi` files from the epytext annotations - possibly from a specifically whitelisted group of modules - so we can work from our existing explicit type annotations and not re-write everything from scratch.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

I'm pretty sure we want type hints inline with code if at all possible.  Maintenance of external stub files, even in the same repository, is a big source of friction.  Do we **need** to use `.pyi` files?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

If we can generate the `.pyi` files from epytext we will get a lot more type coverage more quickly with a lot less diff churn.  It also means we can keep supporting py2 for longer.  That said, if we have the py2 conversation and drop it, inline type hints all the way, yay.  (But then also we want to make pydoctor actually understand type hints, which means porting _it_ to py3, which is a whole separate tooling issue.)

Long term, type annotations is absolutely where I want to end up.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Oh, and the other thing that is important is, as we adopt annotations, we may want to annotate more code internally for type-checking Twisted itself than we publish stubs for initially.  It's easy to accidentally make a signature too strict, or to fail to define a `Protocol` or somesuch and accidentally use a pile of concrete types that are correct internally but wrong for public stubs.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_18"></a>|@wsanchez commented|
|-|-|

pydoctor issue for type hints: https://github.com/twisted/pydoctor/issues/136

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_19"></a>|@mthuurne commented|
|-|-|

The master branch of pydoctor already runs fine on Python 3. That is, I'm using it and not running into any problems. If you encounter any bugs, please report them.

I'm working on implementing annotation support in pydoctor; expect a PR later this week.

I think having a tool to extract type hints from docstrings to `.pyi` files is going to be useful in any case: we can either put the `.pyi` files in a repository or we can apply them to the source code in the form of type comments or annotations. A possible implementation for this would be to add `.pyi` as a report format to pydoctor.

Note that I'm looking at type hints from the perspective of a Twisted user, which is why `.pyi` files are fine for me. It seems the perspective of a Twisted developer is different.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_20"></a>|@wsanchez commented|
|-|-|

Yeah it issue with `.pyi` files is that maintaining them is problematic.  If someone changes code, they may not know to look in a separate file to update a type signature.  This is also why inline documentation is a good thing.

I suggest that (1) you continue the excellent work on getting `pydoctor` to work on Py3 and support type hints (this is incredibly valuable, thanks) and (2) we wait until we know what the end date for Py2 is before we decide on a plan.  I'll be voting for killing Py2 ASAP, as this is but one thing that will be unblocked by dumping it.

**Comment by glyph** [comment]
We already distribute a py.typed and and we have had annotations support in pydoctor for a while now, so while we never distributed stubs per se, the work described here has been done.
